File: gfg/2024_6_june/c++/03_trail_of_ones.cpp
// CODE

class Solution {
  public:
    #define mod 1000000007
    int numberOfConsecutiveOnes(int n) {
        vector<long long> dp(n + 1, 0);
        dp[0] = 0;
        dp[1] = 0;
        dp[2] = 1;
        long long a = 0;
        long long b = 1;
        for(int i = 3; i < n + 1; i++){
            long long c = (a+b) % mod;
            dp[i] = ((dp[i-1] + dp[i-1]) % mod + c % mod) % mod;
            a = b;
            b = c;
        }
        return dp[n];
    }
};

File: gfg/2024_6_june/python/03_trail_of_ones.py
# CODE

class Solution:
    def numberOfConsecutiveOnes(ob, n):
        MOD = 1000000007
        
        dp = [0] * (n + 1)
        dp[0] = 0
        dp[1] = 0
        dp[2] = 1
        
        if(n < 3):
            return dp[n]
        
        a = 0
        b = 1
        for i in range(3, n + 1):
            c = (a + b) % MOD
            dp[i] = ((dp[i - 1] + dp[i - 1]) % MOD + c % MOD) % MOD
            a = b
            b = c
        
        return dp[n]